keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,842 @@
+    namespace that does not own the PID namespace. Pods with `hostPID=true` and
+    `userNamespaceMode=Pod|Cluster` can fail.
+  - Mounting `sysfs` (`/sys`) is not allowed from a process in a user namespace
+    that does not own the network namespace. Impact: pods with
+    `hostNetwork=true` and `userNamespaceMode=Pod|Cluster` can fail.
+
+  If users specify `userNamespaceMode=Pod|Cluster` and one of these
+  `host{IPC,PID,Network}=true` options, runc will currently fail to start the
+  container. The kubelet does **not** try to prevent that combination of
+  options, in case runc or the kernel make it possible in the future to use that
+  combination.
+
+In order to avoid breaking existing workloads `Host` is the default value of
+`userNamespaceMode`.
+
+#### Duplicated Snapshots of Container Images
+
+The owners of the files of a container image have to been set accordingly to the
+ID mappings used for that container. For example, if the user 0 in the container
+is mapped to the host user 100000, then the `/root` directory has to be owned by
+user ID 100000 in the host, so it appears to belong to root in the container.
+The current implementation in container runtimes is to recursively perform a
+`chown` operation over the image snapshot when it's pulled. This presents a risk
+as it potentially increases the time and the storage needed to handle the
+container images.
+
+[containers/storage](https://github.com/containers/storage/) used by CRI-O
+mounts an overlay file system with the
+[`metacopy=on`](https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html#metadata-only-copy-up)
+flag set, it then chowns all of the lower files in the image to match the user
+namespace to which the container will run. This operation is very quick compared
+to standard chowning, since none of the files data has to be copied up. If a
+second container runs on the same image with the same user namespace, then the
+chowned image is shared, eliminating the need to chown again.
+
+[fuse-overlayfs](https://github.com/containers/fuse-overlayfs) is a fuse-based
+overlayfs implementation that supports ID shifting. It doesn't require a
+recursive chown operation (avoiding the file duplication) but given that it's
+implemented in user space has lower I/O performance.
+
+More sophisticated approaches to this problem are being
+[discussed](https://lists.linuxfoundation.org/pipermail/containers/2020-September/042230.html)
+in the kernel community. This is something that container runtimes should use
+once it's available and it does not impact the kubelet nor the CRI gRPC spec.
+
+#### Container Images with High IDs
+
+There are container images designed to run with high user and group IDs. It's
+possible that the IDs range assigned to the pod is not big enough to accommodate
+these IDs, in this case they will be mapped to the `nobody` user in the host.
+
+It's not a big problem in the `Cluster` case, the users have to be sure that
```

Comment:
I'm a little confused about ranges - what are they used for and why are they needed?

Reply:
why do we need ranges for pods?  Why not a single "outside" UID?

Reply:
It's possible that a pod uses multiple IDs inside. It's not possible to map all of those to the same "outside" ID but we need a range. If we only map one ID, then the other IDs will be seen as `nobody - 65534 ` from the host point of view witch creates issues to access files that don't have "others" permissions.

Reply:
@thockin It is implicit on what @mauriciovasquezbernal is saying but the Linux kernel forces you to have a range (you need a 1-1 mapping, a bijection).

I'd like the same you mention (map UIDs/GIDs in the container to just ID X on the host), but it is not possible to do that in Linux.

Reply:
That's simply impossible to do with any guarantee.  Each pod *might* use any UID from the entire space.  More concretely, they could use UID 0 and UID 2^32-1 - requiring the whole range to map.  The only way to guarantee non-collision on the outside is to limit which UIDs can be used on the inside to a "reasonable" range.

This seems like a pretty fundamental thing we need to design for, no?

Sorry if I am rehashing old discussion - I don't see this really covered here.  Are we assuming some "standard" range size for every pod?  Do pods get to express "I need more UIDs"?

One of the very fundamental principles of kubernetes is that users should not be subject to collisions that they did not cause.  For example, sharing host ports.

Hypothetically: Assume we had to pre-allocate "outside" UIDs and the only UIDs you could use were those - would it be better?  What would the unit of allocation-assignment?  I think Namespace or ServiceAccount seems right.  So if we carry that to the hypothetical end, what if we end up at a model like this:

* ServiceAccount gets new "users[]" and "groups[]" fields.  These describe the "inside" and are optional.
* An IDAllocation resource.  This is created in response to the above fields and assigns "outside" IDs to them.
* When a pod runs as that SA it gets that UID mapping.
* When the cluster runs out of IDs, no new IDAllocations can be made.
* Allocation must be pluggable.

That's really the only way to ensure no outside collision.  Is that a worthwhile goal?  Or is the better goal to buy time until node-local ephemeral mappings are viable wrt filesystems?  We still have the questions above - standard range size or configurable, etc.

Reply:
> Sorry if I am rehashing old discussion - I don't see this really covered here. Are we assuming some "standard" range size for every pod? Do pods get to express "I need more UIDs"?

Yes, 65k (UID 0 -> UID 2^16-1) seems to be that [reasonable default range size](https://github.com/kubernetes/enhancements/pull/2101#issuecomment-712960297) (https://github.com/kubernetes/enhancements/pull/2101/files#diff-2d23f433c4d58d94c34afbadc193009e2f7b4b8eab1c6549f7bd96d81abf142bR399-R407). It should be enough for a pod and also allows to schedule up to 65k pods in a node (far away of what k8s handles these days). 

We haven't considered a case where a pod can ask for more UIDs. We'd prefer to see a use case that needs it before proposing something similar.

> One of the very fundamental principles of kubernetes is that users should not be subject to collisions that they did not cause. For example, sharing host ports.

I'm not sure what it means for k8s, but with user namespaces the quantity of IDs available for each pod will be reduced because it's a node-wide shared resource. We'll need to limit the inside range of UIDs a pod can use to something like 0 -> 65k, at least for rev 1. If a pod needs to use IDs outside that range it could share the host user namespace anyway. 

> Hypothetically: Assume we had to pre-allocate "outside" UIDs and the only UIDs you could use were those - would it be better? What would the unit of allocation-assignment? I think Namespace or ServiceAccount seems right. So if we carry that to the hypothetical end, what if we end up at a model like this:

I think you described the allocation algorithm in a high level detail. In general this algorithm should have a pool of host (outside) IDs it can use (maybe configurable by the operator?), each time an "allocation unit resource" (pod / serviceaccount / namespace depending on the outcome of our discussion) is created, this algorithm assigns a portion of that range to that resource.  

> ServiceAccount gets new "users[]" and "groups[]" fields. These describe the "inside" and are optional.

Do you mean to let the user control the "inside" part of the IDs mapping?

> That's really the only way to ensure no outside collision. Is that a worthwhile goal? Or is the better goal to buy time until node-local ephemeral mappings are viable wrt filesystems? We still have the questions above - standard range size or configurable, etc.

That's a good question. What do you mean with "node-local ephemeral mapping"? 
I think we should guarantee that each allocation unit gets a different and non-overlapping range. If we allocate per pod, each kubelet can make this decision, if we allocate per-namespace or per-serviceaccount we'll have to make this decision to an upper layer.

Reply:
>> Sorry if I am rehashing old discussion - I don't see this really covered here. Are we assuming some "standard" range size for every pod? Do pods get to express "I need more UIDs"?
> 
> Yes, 65k (UID 0 -> UID 2^16-1) seems to be that reasonable default range size

Is that really necessary, given that most pods use exactly 1 UID?

> allows to schedule up to 65k pods in a node

Only if you assume every range is localized to a node.  If this was per-SA or
per-NS then it limits you to 64K SAs or NSes, which is actually something
people hit.

What if we gave everyone a smaller set of IDs by default, e.g.
0-1023 + 65520-65535.  Then, anyone who really needed more that that would
have to ask for it, and those special cases could get 64K allocations.

Just brainstorming.

> We haven't considered a case where a pod can ask for more UIDs. We'd prefer to see a use case that needs it before proposing something similar.

I know some apps do run as ID > 64K.  It's not that they need *MORE* IDs,
they'd need *different* ones.

> We'll need to limit the inside range of UIDs a pod can use to something like 0 -> 65k

That's a breaking change for some users.  They need a clean opt-out at least.
That's what the host space is for.  I agree.

>> ServiceAccount gets new "users[]" and "groups[]" fields. These describe the "inside" and are optional.
>
> Do you mean to let the user control the "inside" part of the IDs mapping?

If we end up going as far as an allocator, sure why not?

>> That's really the only way to ensure no outside collision. Is that a worthwhile goal? Or is the better goal to buy time until node-local ephemeral mappings are viable wrt filesystems? We still have the questions above - standard range size or configurable, etc.
> 
> That's a good question. What do you mean with "node-local ephemeral mapping"?

I meant shiftfs or whatever will make FS permissions compatible with user
namespaces.  It sounds like linux is as brain-damaged now as it was when I last
looked at userns - years ago. :(

> I think we should guarantee that each allocation unit gets a different and non-overlapping range. If we allocate per pod, each kubelet can make this decision, if we allocate per-namespace or per-serviceaccount we'll have to make this decision to an upper layer.

I agree with your conclusion, I am less convinced on your predicate.  I am not
sure we *NEED* to *guarantee* that each allocation unit gets a different and
non-overlapping range.  If we simply ensure a low probability of collision, we
can buy enough time for a "real" solution.

I don't know which solution will be "real" enough or how we would migrate
users, but I am sure we can figure that out...

Reply:
> For instance, if a pod gets the `0 100000 65536` mappings and a container inside the pod has `USER: 200000`, it'll be running as `nobody` on the host because the range of length 65536 is not long enough to map this ID to the host.

Ops, I was wrong on this statement. If the user the container should run as is not part of the mapping it'll fail to start. Also calls to `setuid/setgid` within the container trying to set an ID not present on the mapping will fail as well. 

In other words, any container trying to run with an UID not present on the mapping will have a hard failure.

Reply:
> Is that really necessary, given that most pods use exactly 1 UID?

Well, with 64K we’ll cover even more cases. If we go for a per namespace / svc account approach we’ll need to reduce that range as you mentioned above. Finding the right size is important, the bigger the range the less allocations we can do and the higher the probability of a collision, on the other hand, the smaller the range the more applications won’t work. 

> Then, anyone who really needed more that that would
> have to ask for it, and those special cases could get 64K allocations.

I think it’s related to my question about letting the user configure the inside part of the mapping. Probably it’s something we don’t want to do for rev1, asking the user to share the host namespace could be enough for now. For future revisions we could let the user choose the inside part of the mapping, so the IDs they’re using are mapped. 

> I know some apps do run as ID > 64K. It's not that they need _MORE_ IDs,
> they'd need _different_ ones.

Oh yeah, exactly. 

> I meant shiftfs or whatever will make FS permissions compatible with user
> namespaces. It sounds like linux is as brain-damaged now as it was when I last
> looked at userns - years ago. :(

Yes, we don’t have more solutions now. shiftfs will probably never be merged. There is some ongoing work by Christian Brauner[1] but unsure what will happen and, assuming the best case that it is merged and when that kernel will be available in distros

[1]: lwn article: https://lwn.net/Articles/837566/. V4 is out now, some days ago (lwn links to v1 or v2)
 
> I agree with your conclusion, I am less convinced on your predicate. I am not
> sure we _NEED_ to _guarantee_ that each allocation unit gets a different and
> non-overlapping range. If we simply ensure a low probability of collision, we
> can buy enough time for a "real" solution.

I’m not sure either. Having non-overlapping ranges now that we can’t guarantee that each pod runs in a different mapping could be an overkill. I agree we can buy time by having a simpler approach, however I’d like to know the opinion of the Red Hat folks as we know they want to have non-overlapping mappings @mrunalp @giuseppe @rhatdan